### PR TITLE
[RISCV][TypePromotion] Dont generate truncs if PromotedType is greater than Source Type

### LINKED ISSUE #86400: [TypePromotion] Assertion `castIsValid(op, S, Ty) && "Invalid cast!"' failed.
Reduced test case: https://godbolt.org/z/z437vo5MK
```
target datalayout = "e-m:e-p:64:64-i64:64-i128:128-n32:64-S128"
target triple = "riscv64-unknown-linux-gnu"

define i16 @test(i8 %a, i32 %b) {
entry:
  %0 = zext i8 %a to i32
  %conv31.i1161.i = trunc i32 %b to i16
  %cmp.i.i.i1163.i = icmp eq i16 %conv31.i1161.i, 0
  %rem.i.i69.rhs.trunc.i1165.i = trunc i32 %0 to i8
  %1 = zext i8 %rem.i.i69.rhs.trunc.i1165.i to i16
  %2 = xor i16 %1, %conv31.i1161.i
  ret i16 %2
}
```

```
opt: /home/dtcxzyw/WorkSpace/Projects/compilers/llvm-project/llvm/lib/IR/Instructions.cpp:3738: static llvm::CastInst* llvm::CastInst::Create(llvm::Instruction::CastOps, llvm::Value*, llvm::Type*, const llvm::Twine&, llvm::Instruction*): Assertion `castIsValid(op, S, Ty) && "Invalid cast!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: bin/opt -passes=typepromotion test3.ll -S
 #0 0x00007f95a49ec480 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMSupport.so.19.0git+0x1ec480)
 #1 0x00007f95a49e988f llvm::sys::RunSignalHandlers() (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMSupport.so.19.0git+0x1e988f)
 #2 0x00007f95a49e99e5 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007f95a4442520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007f95a44969fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007f95a44969fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x00007f95a44969fc pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x00007f95a4442476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x00007f95a44287f3 abort ./stdlib/abort.c:81:7
 #9 0x00007f95a442871b _nl_load_domain ./intl/loadmsgcat.c:1177:9
#10 0x00007f95a4439e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#11 0x00007f959f298e68 (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.19.0git+0x298e68)
#12 0x00007f95a19391b8 (anonymous namespace)::IRPromoter::Mutate() TypePromotion.cpp:0:0
#13 0x00007f95a193bdab (anonymous namespace)::TypePromotionImpl::TryToPromote(llvm::Value*, unsigned int, llvm::LoopInfo const&) TypePromotion.cpp:0:0
#14 0x00007f95a193c9b5 (anonymous namespace)::TypePromotionImpl::run(llvm::Function&, llvm::TargetMachine const*, llvm::TargetTransformInfo const&, llvm::LoopInfo const&) (.part.0) TypePromotion.cpp:0:0
#15 0x00007f95a193dab4 llvm::TypePromotionPass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCodeGen.so.19.0git+0x73dab4)
#16 0x00007f95a1e6dff6 llvm::detail::PassModel<llvm::Function, llvm::TypePromotionPass, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMPasses.so.19.0git+0x6dff6)
#17 0x00007f959f31cf11 llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.19.0git+0x31cf11)
#18 0x00007f95a3ca55d6 llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMX86CodeGen.so.19.0git+0xa55d6)
#19 0x00007f959f31bc0b llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.19.0git+0x31bc0b)
#20 0x00007f95a3ca69c6 llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMX86CodeGen.so.19.0git+0xa69c6)
#21 0x00007f959f319a81 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.19.0git+0x319a81)
#22 0x00007f95a4d174c5 llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMOptDriver.so.19.0git+0x234c5)
#23 0x00007f95a4d22096 optMain (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMOptDriver.so.19.0git+0x2e096)
#24 0x00007f95a4429d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#25 0x00007f95a4429e40 call_init ./csu/../csu/libc-start.c:128:20
#26 0x00007f95a4429e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#27 0x0000555e00ce9095 _start (bin/opt+0x1095)
Aborted (core dumped)
```

llvm version: b68e2eba0bc8dd70b88f4271831139ee9b6ed25c

cc @topperc

### PR DESCRIPTION
**Pull request by svs-quic**
We currently check if the source and promoted types are not equal before generating truncate instructions. This does not work for RV64 where the promoted type is i64 and this lead to a crash due to the generation of truncate instructions from i32 to i64.

Fixes #86400

### CODE REVIEW

**Comment by svs-quic** [comment]
@topperc @dtcxzyw

File: llvm/test/CodeGen/RISCV/rv64-typepromotion.ll
Comment (dtcxzyw) [inline llvm/test/CodeGen/RISCV/rv64-typepromotion.ll:21]: Can you simplify these value names?

**Review by wangpc-pp — APPROVED** [review]
LGTM.

**Comment by svs-quic** [comment]
Could someone please merge this if everyone is okay with the change? I don't have commit access.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**
